lib/pos_blockchain.go

```diff
@@ -0,0 +1,1331 @@
+package lib
+
+import (
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/deso-protocol/core/collections"
+	"github.com/deso-protocol/core/consensus"
+	"github.com/dgraph-io/badger/v3"
+	"github.com/golang/glog"
+	"github.com/pkg/errors"
+)
+
+// processBlockPoS runs the Fast-Hotstuff block connect and commit rule as follows:
+//  1. Determine if we're missing the parent block of this block.
+//     If so, return the hash of the missing block and add this block to the orphans list.
+//  2. Validate the incoming block, its header, its block height, the leader, and its QCs (vote or timeout)
+//  3. Store the block in the block index and save to DB.
+//  4. try to apply the incoming block as the tip (performing reorgs as necessary). If it can't be applied, we exit here.
+//  5. Run the commit rule - If applicable, flushes the incoming block's grandparent to the DB
+func (bc *Blockchain) processBlockPoS(block *MsgDeSoBlock, currentView uint64, verifySignatures bool) (
+	_success bool, _isOrphan bool, _missingBlockHashes []*BlockHash, _err error) {
+	// If we can't hash the block, we can never store in the block index and we should throw it out immediately.
+	if _, err := block.Hash(); err != nil {
+		return false, false, nil, errors.Wrapf(err, "processBlockPoS: Problem hashing block")
+	}
```

Comment:
In another comment I mention breaking the block sanity check and the QC check out into something like `validateBlockQC()`. If you do that, I would put that function right here, and I would return an error without storing anything if it fails. This would allow the caller to disconnect the node that sent the messed up block without opening us to spam possibillity. It would also allow you to store invalid blocks in later sections of the code, like in the next if statement, which is useful.